[TexMap] Improve painting tile fragments in compositor

Scony opened the pull request:
#### 3d4e69f7f44d9a782396282d956f4ab99ac5773a
<pre>
[TexMap] Improve painting tile fragments in compositor
<a href="https://bugs.webkit.org/show_bug.cgi?id=290169">https://bugs.webkit.org/show_bug.cgi?id=290169</a>

Reviewed by Carlos Garcia Campos.

This change utilizes the Damage to further improve
the tile painting in the compositor.

If tile size allows, with this change it&apos;s now possible
to have multiple &quot;draw&quot; operations per tile.

* Source/WebCore/platform/graphics/Damage.h:
(WebCore::Damage::Damage):
(WebCore::Damage::rectsForPainting const):
(WebCore::Damage::makeFull):
(WebCore::Damage::add):
(WebCore::Damage::initialize):
(WebCore::Damage::shouldAdd const):
(WebCore::Damage::tileIndexForRect const):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
(WebCore::CoordinatedBackingStore::paintToTextureMapper):
* Tools/TestWebKitAPI/Tests/WebCore/glib/Damage.cpp:
(TestWebKitAPI::TEST(Damage, Unite)):
(TestWebKitAPI::TEST(Damage, RectsForPainting)):

Canonical link: <a href="https://commits.webkit.org/292700@main">https://commits.webkit.org/292700@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e4c66b0e70bea1684909c3565635ddb0704c98b7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96839 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16453 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6631 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101912 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47359 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98884 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16749 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24897 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73791 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31001 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99842 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12612 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87593 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54125 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12372 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5410 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46687 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/82466 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5490 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103935 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23907 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17441 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82838 "Passed tests") | 
| | [❌ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24158 "Failed to checkout and rebase branch from PR 42805") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83664 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82228 "Found 3 new API test failures: /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print-errors, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/close-after-print (failure)") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4436 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17398 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15603 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23869 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29024 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23528 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27008 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25269 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -76,12 +81,13 @@ class Damage {
     ALWAYS_INLINE bool isEmpty() const  { return m_rects.isEmpty(); }
 
+    // FIXME: Pass rect.
     void makeFull(const IntSize& size)
```
Let's do it in this patch, it shouldn't be  much code.

carlosgcampos commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -44,22 +44,27 @@ class Damage {
     };
 
-    static constexpr int s_tileSize { 256 };
+    static constexpr int s_cellSize { 256 };
```
I made this public and added `m_tileSize` becase we were not yet suyre whether to allow passing that as constructor. I think passing the size with a fixed tile/cell size is working fine, so we can make this private and remove the member

carlosgcampos commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -44,22 +44,27 @@ class Damage {
     };
 
-    explicit Damage(const IntSize& size, Mode mode = Mode::Rectangles)
+    explicit Damage(const IntRect& gridRect, Mode mode = Mode::Rectangles)
```
I would call this just rect, grid is an implementation detail and it's only used in some cases.

carlosgcampos commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -161,26 +167,52 @@ class Damage {
+        if (m_mode != Mode::Rectangles)
+            return;
+
+        Rects rectsCopy = m_rects;
```
Since we are copying here, we can take advantage to remove the empty rects and then we don't need to worry about then while iterating below.

carlosgcampos replied:
```suggestion
        auto rectsCopy = WTF::compactMap(m_rects, [](const auto& value) -> std::optional<IntRect> {
            if (value.isEmpty())
                return std::nullopt;
            return value;
        });
```

carlosgcampos commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -161,26 +167,52 @@ class Damage {
             add(rect);
     }
 
+    void removeOverlapsWhileClippingToGrid()
```
Since this is always called before iterating the rects for painting, I would make this a getter instead of modifying the damage. For non-rectangles case I will always copy, but it's always just one rect.

carlosgcampos commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -161,26 +167,52 @@ class Damage {
+                IntRect minimumBoundingRectangleContaingOverlaps;
+                for (const auto& rect : rectsCopy) {
+                    if (rect.isEmpty())
+                        continue;
```
This shouldn't be needed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
```diff
@@ -82,17 +82,26 @@ void CoordinatedBackingStore::paintToTextureMapper(TextureMapper& textureMapper,
+            // area significantly in many cases.
+            const bool shouldUseAtMostOneDrawCall = tile.rect().width() <= Damage::s_cellSize && tile.rect().height() <= Damage::s_cellSize;
+            IntRect tileIntRect = enclosingIntRect(tile.rect());
+            Damage tileDamage(tileIntRect, shouldUseAtMostOneDrawCall ? Damage::Mode::BoundingBox : Damage::Mode::Rectangles);
```
I think we can optimize the case of using rectangles when the area is smaller than cell size to be equivalent to bounding box mode. We could check in unite() if the vector size is 1 and then just set the minimum bounding box at 0.

Scony replied:
I just noticed `tileIndexForRect` early returns 0 if `m_rects.size() == 1` so we're good

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
```diff
@@ -82,17 +82,26 @@ void CoordinatedBackingStore::paintToTextureMapper(TextureMapper& textureMapper,
+                IntRect tileDamageRect = intersection(tileIntRect, damageRect);
+                if (tileDamageRect.isEmpty())
+                    continue;
+                tileDamage.add(tileDamageRect);
```
Damage::add already returns early if the rectangle is empty, so this could be simplified and just do:

```
tileDamage.add(intersection(tileIntRect, damageRect));
```

carlosgcampos commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -166,21 +195,21 @@ class Damage {
     ALWAYS_INLINE bool shouldAdd() const
     {
-        return !m_size.isEmpty() && m_mode != Mode::Full;
+        return !m_rect.size().isEmpty() && m_mode != Mode::Full;
```
```suggestion
        return !m_rect.isEmpty() && m_mode != Mode::Full;
```

Scony committed:
[TexMap] Improve painting tile fragments in compositor
https://bugs.webkit.org/show_bug.cgi?id=290169

Reviewed by Carlos Garcia Campos.

This change utilizes the Damage to further improve
the tile painting in the compositor.

If tile size allows, with this change it's now possible
to have multiple "draw" operations per tile.

* Source/WebCore/platform/graphics/Damage.h:
(WebCore::Damage::Damage):
(WebCore::Damage::rectsForPainting const):
(WebCore::Damage::makeFull):
(WebCore::Damage::add):
(WebCore::Damage::initialize):
(WebCore::Damage::shouldAdd const):
(WebCore::Damage::tileIndexForRect const):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
(WebCore::CoordinatedBackingStore::paintToTextureMapper):
* Tools/TestWebKitAPI/Tests/WebCore/glib/Damage.cpp:
(TestWebKitAPI::TEST(Damage, Unite)):
(TestWebKitAPI::TEST(Damage, RectsForPainting)):

Canonical link: https://commits.webkit.org/292700@main

nikolaszimmermann commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -76,25 +78,52 @@ class Damage {
+            return m_rects;
+
+        Rects rects;
+        for (int row = 0; row < m_gridCells.height(); row++) {
```
s/row++/++row/

nikolaszimmermann commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -76,25 +78,52 @@ class Damage {
+        Rects rects;
+        for (int row = 0; row < m_gridCells.height(); row++) {
+            for (int col = 0; col < m_gridCells.width(); col++) {
+                IntRect tileRect = { { m_rect.x() + col * s_cellSize.width(), m_rect.y() + row * s_cellSize.width() }, IntSize { s_cellSize.width(), s_cellSize.width() } };
```
You can directly use s_cellSize in the last argument.

nikolaszimmermann commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -166,21 +195,21 @@ class Damage {
         case Mode::Rectangles:
-            m_gridSize = ceiledIntSize({ static_cast<float>(m_size.width()) / m_tileSize.width(), static_cast<float>(m_size.height()) / m_tileSize.height() }).expandedTo({ 1, 1 });
-            m_shouldUnite = m_gridSize.width() == 1 && m_gridSize.height() == 1;
+            m_gridCells = IntSize(std::ceil(static_cast<float>(m_rect.size().width()) / s_cellSize.width()), std::ceil(static_cast<float>(m_rect.size().height()) / s_cellSize.height())).expandedTo({ 1, 1 });
```
why did you stop using ceiledIntSize?

Scony replied:
`ceiledIntSize` lead to the following:
`(1.003906,0.195312) -> m_gridCells:(1,1)` -  this is because when using `ceiledIntSize` we create `LayoutSize` from the `IntSize` and `LayoutSize` uses `LayoutUnit` for storing and that in turn uses `int` as a storage. So we effectively store `float` using `int` (`LayoutUnit`) thus loosing precision, and then we ceil, and get `IntSize` back

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
```diff
@@ -82,17 +82,21 @@ void CoordinatedBackingStore::paintToTextureMapper(TextureMapper& textureMapper,
+            // If the tile size is big enough, the Damage will have a better resolution
+            // and will allow more than one "draw" operation thus reducing the re-painted
+            // area significantly in many cases.
+            const IntRect tileIntRect = enclosingIntRect(tile.rect());
```
const auto

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
```diff
@@ -82,17 +82,21 @@ void CoordinatedBackingStore::paintToTextureMapper(TextureMapper& textureMapper,
+                tileDamage.add(intersection(tileIntRect, damageRect));
+            }
+            for (const auto& tileDamageRect : tileDamage.rectsForPainting()) {
+                const auto sourceRect = FloatRect { FloatPoint { tileDamageRect.location() - tile.rect().location() }, tileDamageRect.size() };
```
You could have omitted the FloatPoint

Scony replied:
`error: converting to 'const WebCore::FloatPoint' from initializer list would use explicit constructor 'WebCore::FloatPoint::FloatPoint(const WebCore::FloatSize&)'`

nikolaszimmermann reviewed (commented):
Got some post-commit comments, sorry for being late.
